beginner_source/translation_transformer.py

```diff
@@ -303,7 +303,7 @@ def train_epoch(model, optimizer):
         optimizer.step()
         losses += loss.item()
 
-    return losses / len(train_dataloader)
+    return losses / len(list(train_dataloader))
```

Comment:
It might be better to fix the `__len__` of ZipperIterDataPipe in torchdata: https://github.com/pytorch/pytorch/blob/ea72a0991c3422d8f314acdf8b911de42a6b4c1e/torch/utils/data/datapipes/iter/combining.py#L554

cc: @ejguan

Reply:
It's better to append a `LenthSetterDataPipe` at the end of each Dataset within TorchText with a known length. For `Zipper` in torchdata, it's for general usage that requires the prior `DataPipe` provides valid length.

Reply:
cc: @Nayef211

Reply:
@ejguan if my understanding is correct, calling `list(train_dataloader)` would cause all data in the dataset to materialize right? I thought this is something we'd want to avoid given the fact that this can lead to OOMs for very large datasets.

> It's better to append a LenthSetterDataPipe at the end of each Dataset within TorchText with a known length.

If we do this, can we avoid the data materialization issue I mentioned above?

Reply:
Correct, we shouldn't do materialization here. If this Dataset has a known length, we can add a DataPipe:
```py
class LenSetter(IterDataPipe):
    def __init__(self, length):
        self.length = length

    def __len__(self):
        self.length = lenth
```
This would prevent any pre-materialization before data iterated. `IterDataPipe` itself doesn't guarantee to provide a `__len__` function because there might DataPipe operations introduce the dynamic length (`filter`)

Reply:
Gotcha, just created an issue to track this https://github.com/pytorch/text/issues/1943. Will merge this PR for now